File: Java/n-queens.java
//  N-Queens

// Time O(N*N!), Space O(N)
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> ans = new ArrayList<List<String>>();
        int[] C = new int[n];
        for (int i=0; i<n; i++)
            C[i] = -1;
        dfs(n, 0, C, ans);
        return ans;
    }
    
    private void dfs(int n, int row, int[] C, List<List<String>> ans) {
        if (row == n) {
            List<String> l = new ArrayList<String>();
            for (int i=0; i<n; i++) {
                char[] ch = new char[n];
                Arrays.fill(ch, '.');
                ch[C[i]] = 'Q';
                l.add(new String(ch));
            }
            ans.add(l);
            return;
        }
        for (int col=0; col<n; col++) {
            if (!isValid(C, row, col))
                continue;
            C[row] = col;
            dfs(n, row+1, C, ans);
            C[row] = -1;
        }
    }
    
    private Boolean isValid(int[] C, int row, int col) {
        for (int i=0; i<row; i++) 
            if (col==C[i] || Math.abs(i-row)==Math.abs(C[i]-col))
                return false;
        return true;
    }
}

File: Leetcode200/n-queens.cc
//  N-Queens

class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> res;
        vector<int> flag(n, -1);
        dfs(n, 0, flag, res);
        return res;
    }
    
    void dfs(int n, int cur, vector<int> &flag, vector<vector<string>> &res) {
        if (cur == n) {
            vector<string> path;
            for (int i=0; i<n; i++) {
                string s(n, '.');
                for (int j=0; j<n; j++)
                    if (flag[i] == j) { s[j] = 'Q'; break; }
                path.push_back(s);
            }
            res.push_back(path);
            return;
        }
        for (int col=0; col<n; col++) {
            if (!is_valid(cur, col, flag))  continue;
            flag[cur] = col;
            dfs(n, cur+1, flag, res);
            flag[cur] = -1;
        }
    }
    
    bool is_valid(int row, int col, vector<int> &flag) {
        for (int i=0; i<row; i++) 
            if (col==flag[i] || abs(col-flag[i])==abs(row-i))
                return false;
        return true;
    }
};